### DISCUSSION TITLE
Add: allow ref this parameter in extension methods for variables created implicitly by the compiler

### DISCUSSION THREAD

**Discussion by marchewek** [dotnet/csharplang#8324] [Language Ideas]
First things first to set the context - I find myself very good in object-orientation paradigm and consistency (as a kind of _life philosophy_), but I never excelled in the low level stuff (main reason I decided to stick with C# instead of C++). Therefore my finding doesn't have to necessarily be a "missing language feature" but simply a "documentation about that low level stuff in C# is poor"-kind-of finding. Please keep that in mind going forward. I'm also not an English-native, so please forgive me any linguistic mistakes I made along the way. This is a follow up to [the original post I made in the incorrect repo*](https://github.com/dotnet/runtime/issues/105252).

As pointed out in my original post, the language right now behaves differently for explicit and implicit variables (by _implicit_ I mean a variable created by the compiler in the background):
``` csharp
public class TestCode
{
	internal static void Main()
	{
		// with explicit variable it works well
		var x = new SomeStruct();
		x.ExtensionsRefParam();
		x.ExtensionRefParamRefRet();

		// with implicit variable neither compiles (triggers CS1510)
		new SomeStruct().ExtensionsRefParam();
		new SomeStruct().ExtensionRefParamRefRet();

		// compiles, but copies the struct, so out of scope; kept for completeness 
		new SomeStruct().InstanceMethod();
		new SomeStruct().ExtensionNoRefParam();
	}
}

public struct SomeStruct
{
	public SomeStruct InstanceMethod()
	{
		return this;
	}
}

public static class SomeExtensions
{
	public static SomeStruct ExtensionNoRefParam(this SomeStruct @this)
	{
		return @this;
	}

	public static SomeStruct ExtensionsRefParam(ref this SomeStruct @this)
	{
		return @this;
	}

	public static ref SomeStruct ExtensionRefParamRefRet(ref this SomeStruct @this)
	{
		return ref @this;
	}
}
```

Now as @stephentoub pointed out in the original post, I could use the "in" indicator for the argument and in my case that actually solved the problem - I never intended to modify the implicit variable. But what if someone actually wants to modify the variable? Why make it inconsistent between an explicit and implicit variable?

Once again, my finding comes mostly from _consistency_ standpoint; if this is something impossible due to technical constraints or I am asking for something that would enable some explicitly-unintended scenarios then maybe improving documentation about this part of language would be good enough.

[EDIT 1]
The OP lacked particular use cases, so here are the two that emerged from the discussion.

1. Implementing the fluent builder pattern via extension methods while at the same time not wanting to copy the struct in (ref parameter) nor out (return ref), because the structure is big:
``` csharp
var built = new SomeStruct().Config1().Config2().Config3().Build();
```

2. Implementing the fluent _configuration_ pattern while [... the same as above]:
``` csharp
var someStruct = ref new SomeStruct().Config1().Config2().Config3();
```
Personally I find `ref` here redundant (makes the syntax more complex), but it may be required due to technical constraints (makes the language definition less complex).

\* sorry about that; those are my baby steps in contributing to C#

**Comment by HaloFour** [comment]
I mean, it is consistent in that you can't pass that expression to a method that accepts a `ref` parameter either.  But beyond that, doing something for the sake of consistency doesn't seem to be all that important if the feature doesn't provide value by itself.

**Reply by marchewek** [reply]
> I mean, it is consistent in that you can't pass that expression to a method that accepts a `ref` parameter either.

Nope, you actually CAN pass the x to the `x.ExtensionsRefParam();` (which it is ref parameter).

> doing something for the sake of consistency doesn't seem to be all that important 

While I agree that consistency on its own shouldn't be the goal, I still find it a bit confusing that some variables (explicitly defined by the programmer) are treated differently than some others (implicitly created by the compiler).

**Reply by CyrusNajmabadi** [reply]
it's unclear why you want this to work.  What's the actual code you're trying to write where this would be a good thing?

**Reply by marchewek** [reply]
My question is on a different abstraction level; I'm not asking "how to implement something", but rather "why the language allows a specific construct for one variable (defined by the programmer) while throwing compilation error for another one (created by the compiler in the background)" - is it intentional (in order to avoid some misuses) or unintentional (ooops we didn't consider this scenario)?

**Reply by CyrusNajmabadi** [reply]
> but rather "why the language allows a specific construct 

Because we think that construct makes sense and has use cases worth supporting. 

> while throwing compilation error for another one

Because there aren't use cases for this. That's why I asked what you needed this for :-)

If something doesn't have use cases, why would we add support for it?

**Reply by marchewek** [reply]
I don't understand why you are asking for any specific implementation - conceptually the use cases for `ref this` are the same regardless of how the variable is created. If you are saying that there is no value in supporting variables created by the compiler, then there is no value in supporting `ref this` at all.

In sarcastic words my use case is "as a programmer I want to save some lines of code, because I'm not paid for kloc".

In less sarcastic words here's another example:
``` csharp
public ref struct MyStruct // defined somewhere
{
	public string Prop;
}

public static class Extensions
{
	// defined somewhere else, hence cannot be the struct's method
	public static ref MyStruct ComputeComputeCompute(ref this MyStruct ms)
	// I don't want to copy MyStruct (hence "ref") and I do want to change it (hence not "in")
	{
		ms.Prop = "we do not support variables created by the compiler, ha ha ha!"; 
		return ref ms;
	}
}

internal class Program
{
	public static void Main()
	{
		// possible
		var x = new MyStruct();
		var y = x.ComputeComputeCompute().Prop;

		// not possible
		var z = new MyStruct().ComputeComputeCompute().Prop;

		// not possible either
		var zztop = SomeMethodReturningMyStruct().ComputeComputeCompute().Prop;
}
```

If you are saying that when you implemented this feature you intentionally added the support for the variables explicitly declared by the programmer while intentionally not adding the support for the variables created by the compiler, then I was simply interested in the rationale for the latter...

**Reply by CyrusNajmabadi** [reply]
> I don't understand why you are asking for any

Language features need compelling use cases. If those don't exist, we don't work to make something work.

**Reply by marchewek** [reply]
> conceptually the use cases for `ref this` are the same regardless of how the variable is created

**Reply by CyrusNajmabadi** [reply]
I don't understand the case. Ref would be to modify the caller. But you haven't passed a location that can be modified. Again, what is a real case where you need this.  

Given that you haven't shown any cases where you would use this, I think that helps explain why we haven't added support for it.

**Reply by marchewek** [reply]
> you haven't passed a location that can be modified

Both `x` and compiler-generated storage on stack for `new MyStruct()` can be modified...

One last try, though I'm going in circles (`ref` is now removed from struct declaration `public struct MyStruct`):
``` csharp
public struct MyStruct // defined somewhere
{
	public string Prop;

	public MyStruct ComputeComputeInstanceMethod()
	{
		Prop = "computed";
		return this;
	}
}

public static class Extensions
{
	// defined somewhere else, hence cannot be the struct's method
	public static ref MyStruct ComputeComputeCompute(ref this MyStruct ms)
		// I don't want to copy MyStruct (hence "ref") and I do want to change it (hence not "in")
	{
		ms.Prop = "we do not support variables created by the compiler, ha ha ha!";
		return ref ms;
	}
}

internal class Program
{
	public static void Main()
	{
		// ofc possible with instance method
		var z = new MyStruct() { Prop = "123" }.ComputeComputeInstanceMethod().Prop;
		Console.WriteLine(z); // displays: computed

		// possible with extension method on programmer-defined variable
		var x = new MyStruct() { Prop = "123" };
		var y = x.ComputeComputeCompute().Prop;
		Console.WriteLine(y); // displays: we do not support variables created by the compiler, ha ha ha!

		// not possible with extension method on compiler-generated variable
		var zzz = new MyStruct() { Prop = "123" }.ComputeComputeCompute().Prop;
	}
}
```

Please note that semantically the "not possible with extension method on compiler-generated variable" is the exact same as "possible with extension method on programmer-defined variable" - it's just compacting two lines into one.

**Reply by CyrusNajmabadi** [reply]
> can be modified...

Not in a meaningful way.  The user can't reference that location to observe anything.  In compiler/language parlance, this is an rvalue, not an lvalue.

**Reply by marchewek** [reply]
> can't reference that location to observe anything

well, let's make a small change to that example, so that the `struct` (not its property) is referenced by the user:

``` csharp
var x = new MyStruct() { Prop = "123" };
var y = x.ComputeComputeCompute();
// possible, but doesn't really make much sense
// x.ComputeComputeCompute() and using x onward would be enough

var zzz = new MyStruct() { Prop = "123" }.ComputeComputeCompute();
// impossible, however it would make sense
// we could be operating on a struct modified by the extension method already from the start
// in some of the previous scenarios (without using the initializer) it would be more readable than its two-liner equivalent
```

**Reply by HaloFour** [reply]
This feels like it's going in circles.  The feature wasn't implemented this way because it wasn't found necessary to do so.  The team doesn't do additional work for the sake of consistency.  If you are interested in this feature being supported, providing actual use cases would help the case.

**Reply by CyrusNajmabadi** [reply]
> new MyStruct() { Prop = "123" }.ComputeComputeCompute();

You don't need 'ref' for this.  'ref' indicates you want to mutate that storage location, which is not observable here (since it is an rvalue not an lvalue).  If you aren't mutating the storage location, you can use `in`.  And if you're doing this just for the return value, you don't need to mutate the storage location.

Again, none of these cases demonstrate why we'd want to support this case.

**Reply by CyrusNajmabadi** [reply]
> The team doesn't do additional work for the sake of consistency.

Also, there's no questoin of consistency here.  We are consistent.  We support these features on lvalues, not on rvalues.  That's consistent in not wanting to offer features that imply unobservable results.  If  we were to add support for this, it would have to be with a deeply good case demonstrating a critical coding pattern that is currently blocked that should be allowed.  

Literally no real cases have been provided.  And all the synthetic cases further reinforce why we would not think this was a good idea.

**Reply by marchewek** [reply]
``` csharp
public static class Extensions
{
	public static void Cfg1(ref this MyStruct ms) { }
	public static void Cfg2(ref this MyStruct ms) { }
	public static void Cfg3(ref this MyStruct ms) { }

	public static ref MyStruct Config1(ref this MyStruct ms) => ref ms;
	public static ref MyStruct Config2(ref this MyStruct ms) => ref ms;
	public static ref MyStruct Config3(ref this MyStruct ms) => ref ms;
}

internal class Program
{
	public static void Main()
	{
		// so 00s...
		var x = new MyStruct();
		x.Cfg1();
		x.Cfg2();
		x.Cfg3();

		// yay, fluent API... kinda... because...
		var y = new MyStruct();
		y.Config1()
			.Config2()
			.Config3();
			
		// ...CS1510; go ask your provider to change MyStruct to a class, problem solved
		var z = new MyStruct()
			.Config1()
			.Config2()
			.Config3();
	}
}
```

**Reply by jilles-sg** [reply]
> not wanting to offer features that imply unobservable results

Following that reasoning, it's actually instance methods on mutable structs that are the odd one out, not `ref this` extension methods.

This is quite insidious: something like `structs[0].Mutate()` will change or leave unchanged `structs[0]` depending on what `structs` is (for example, it's changed if `structs` is `SomeStruct[]` or `Span<SomeStruct>`, but unchanged if `structs` is `List<SomeStruct>` or `ReadOnlySpan<SomeStruct>`).

On the other hand, a call to a `ref this` extension method will safely fail to compile if the value cannot be mutated. In the case of the `List<SomeStruct>`, the programmer can then read the value out of the list, mutate it, and store it back; this is annoying, but it doesn't cause unexpected behaviour.

If the programmer writes `SomeStruct s = structs[0]; s.Mutate();`, they are assumed to understand that this will not affect the value in `structs`.

The explanation is probably historical, since `readonly structs` and `readonly` methods did not use to exist, and, for example, given a `List<(int a, int b)>` or a `List<T> where T: IComparable<T>` it is desirable to be able to call the `CompareTo` method on the elements without needing to copy explicitly.

**Reply by marchewek** [reply]
okay, it's consistent, whatever...

**Reply by CyrusNajmabadi** [reply]
> okay, it's consistent, whatever...

Look, i'm not certain how many more ways it can be stated.  There were reasons we did it this way.  And we'll change if we get a compelling reason to do so.  I'm not sure what else you would want here.

**Reply by CyrusNajmabadi** [reply]
> // yay, fluent API... kinda... because...

Why is your fluent api mutating an rvalue in place?  Fluent APIs normally return a final value, with new intermediary non-mutated rvalues in situ.  

You haven't explained why your fluent api works by mutating hte actual rvalue.

**Reply by snarbies** [reply]
It's not unusual for a function in a fluent API to serve the express purpose of mutating the underlying object (think jquery). 

It doesn't seem absurd to want to create and mutate an object, in order to configure it, and then have it produce a value, either without ever actually needing to explicitly store a reference to that object or with the value produced being the original object itself.

```csharp
		var z = new MyStruct()
			.Config1()
			.Config2()
			.Config3();
```
In this case, my expectation would be that the struct will live on the stack, will be mutated by the "Config" methods, and finally produce a value (itself). 

If, in practice, each of these method calls operates on a new, different temporary, and the result isn't what you expect, then prohibiting this kind of call is a good thing. 

If not, it strikes me as an arbitrary restriction.

It's not clear to me whether these ref extension methods are returning a reference to the original "rvalue" or.... what? But the *desired* functionality seems reasonable.

**Reply by CyrusNajmabadi** [reply]
> It's not unusual for a function in a fluent API to serve the express purpose of mutating the underlying object (think jquery).

Can you point to such an API in .net?  Or a request for a particular API to work that way? 

jQuery doesn't apply. It's not a stack only API.  The equiv in .net wouldn't have any issues. 

> But the desired functionality seems reasonable.

The use cases don't seem normal or expected to me.  Especially the mutating in place then returning a copy.  Not only is it strange, it seems super error prone. If you capture a value midway, but then keep fluently mutating, you could end up with all sorts of strange results.  As opposed to how fluent apis normally work with clear chaining semantics.

**Reply by snarbies** [reply]
> Can you point to such an API in .net? Or a request for a particular API to work that way?

Nope. I don't have a dog in this fight, and frankly the programming I do doesn't involve tons of external libraries, but the lack of a handy example at hand doesn't render the idea invalid. It's something I might do myself if it were possible (and if I saw a potential benefit), for whatever that's worth.

> The use cases don't seem normal or expected to me.

I really don't know what to say to that. If it's a coherent solution, I don't know what's not "normal" or "expected". If it's a value type you shouldn't be surprised that it acts like a value type. But I personally feel a lot less, uh... challenged... by mutable value types than most people seem to be. 

> If you capture a value midway

It seems weird and unexpected to do that with a fluent API, defeating the purpose. But if you did, that could even be construed as a feature, "forking" the process. But mostly I'd just refer back to "value types act like value types". I'm just... not seeing "super error-prone".

It almost feels like you don't get it because it's not something one would do because it's not something one *could* do because it was disallowed because you don't get it because it's not something one would do because it's not something one *could* do because it was disallowed... It's a reasonable, coherent, presumably workable idea that you just don't seem to like the taste of. 

Like I said, I don't have a dog in this fight. I just see two people talking past each other and tried to help bridge that gap. For the sake of my own curiosity, is there any *technical* challenge here? Equally important, would there be any kind of *benefit*, performance or otherwise, if this were allowed? I assume that the point of doing this via refs is performance, or else why are we spinning our wheels?

**Reply by marchewek** [reply]
I stumbled upon this when benchmarking performance of different implementations of a certain thing and found it weird that it's disallowed. I was more into fluent builder pattern `new MyStruct().Config1().Config2().Config3().Build();` rather than getting back MyStruct itself (maybe in latter it should be something like `var zzz = ref new MyStruct().Config1().Config2();` but that use case emerged during the discussion).

>  you don't get it because it's not something one would do because [...]

I like that one :) on a more serious note - yes I don't get _why_ it's disallowed and I have the right to not know all the rationale for all the decisions because I neither work in C# language team (and even then I'd argue that I wouldn't have to know rationale for every single decision) nor I've read the entire Internet (I have limited time to read everything).

> There were reasons we did it this way. 

Simply responding "listen we did it this way because this and that" would be a perfectly valid answer to my post. It's just that I proposed an idea and I'm getting "we won't tell you why we did it this way unless you prove us wrong and then we will reconsider". After all, I wrote...

> my finding doesn't have to necessarily be a "missing language feature"

**Reply by marchewek** [reply]
Updated the OP with example use cases.

**Reply by CyrusNajmabadi** [reply]
> we won't tell you why we did it this way unless you prove us wrong

I think you have a misunderstanding of lang design. We add support for something when there is a *strong* reason to do so.  

There doesn't need to be a justification to not do something. That's the default state/position of the lang.  If you want something, an extremely good case needs to be made *why*. 

So far, that case has not been made.

**Reply by marchewek** [reply]
> I think you have a misunderstanding of lang design

And I think we just use different architecture documentation standards. When I implement anything that is shared I document which scenarios I decided to address and which I decided not to address, both with some rationale.

> So far, that case has not been made.

For me it's a natural extension of "we allow you to reduce the number of heap allocations (use structures instead of classes) and avoid copying them in case they are large (with ref-in and ref-out)".

I did provide two use cases and updated the post, what should be enough to showcase the intended scenario. If you don't like that idea, then I guess fine, you don't have to, and feel free to close the discussion.

**Reply by jilles-sg** [reply]
You can do something like this today:
```csharp
using System.Diagnostics.CodeAnalysis;

struct SomeStruct {
    [UnscopedRef]
    public ref SomeStruct StartBuilding() => ref this;
}
```

Once `StartBuilding` has been called, it's also possible to call extension methods.

The nice thing about mutating the receiver and returning `ref` (apart from performance considerations) is that the return value can be ignored, so both
```csharp
var result = new SomeStruct().StartBuilding().Config1().Config2().Build();
```
and
```csharp
SomeStruct builder = new();
builder.Config1();
builder.Config2();
var result = builder.Build();
```
work.

I kind of like the idea, but without move-only/non-copyable structs it is too easy to mutate a copy accidentally. So I don't think the first step here should be changing the behaviour of `ref this` extension methods.

**Reply by CyrusNajmabadi** [reply]
> . When I implement anything that is shared I document which scenarios I decided to address and which I decided not to address, both with some rationale.

We have given our rationale. Specifically that we don't add support for language features that no compelling use case has been made for. 

> I did provide two use cases and updated the post, what should be enough to showcase the intended scenario. 

The op does not do this. It demonstrates the feature, but gives no reason or example for why it's a good thing. It's telling that the name of the struct is simply "SomeStruct". Something vague without any actual example showing a use case where this would be wanted. 

> If you don't like that idea, 

I don't have any opinion on the idea yet.  All I've said is that we need compelling use cases to consider language support and that no good cases have been presented. And even the vague things that have been alluded to come with fairly substantial problems that we would be wary about.

**Reply by marchewek** [reply]
@jilles-sg thanks for the idea 👍 it requires some modification of the original structure though.

@CyrusNajmabadi sorry for not meeting your demand to provide a very concrete use case; it's because I don't have one. My line of thought was more on a conceptual level - it was like "here you have reference types, but meh it's heap allocation, garbage collection takes time and stuff, so we want to allow you to write some more efficient code by using stack allocations instead, so we added ref in (a long time ago) and a ref out (recently), but you know what? no, you won't be able to write the same API on classes and ref structs fully avoiding copying (in other words - you won't have the same developer experience), though conceptually both are pointers to memory (just in a different location)" and I thought like "well, this is a bit surprising", because I was expecting to be able to get similar behavior for ref structs that I had for classes and instead I got a compiler error.

That said, it doesn't mean that I will suddenly switch to java (as we know they don't see sharp; sorry it was stronger than me).

> I don't have any opinion on the idea yet.

Feel free to close the discussion if you come to a conclusion that you dislike it. I've already shared all my thoughts.

P.S. Yes, I will be missing details of defining the language - I wrote a compiler only once in my life, during my studies, some 15 years ago, and concluded that it's not my kind of thing. Luckily there are people who like this kind of stuff.

**Reply by CyrusNajmabadi** [reply]
> sorry for not meeting your demand to provide a very concrete use case; it's because I don't have one

No worries. 

Absent a use case, our default position is: we will not add support for it. 

We have limited resources and tons of things that will be helpful, so we spend our efforts there. 

> When I implement anything that is shared I document which scenarios I decided to address and which I decided not to address

I wanted to address this.  There are a finite set of things we have added support for, and an infinite number we haven't.  We can't document the reasons for the infinite set of things as that would be an unending process :-)

**Comment by FaustVX** [comment]
You can use the `in` keyword instead of `ref`.

You just have to use `ref readonly` return on the last extension

[sharplab.io](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+ABAzAAmwCZ8AVCAZ2AGEB7FCAWACgBvFgSBgDthpuAhgBtCARgBshACz4AsgJ4AKAJSd2zDhwD0W/AHcYwABb4IADwAOQmAGND+MANgCARkIj57e2lADWFfQghIU4ORyh8M3wAXnxuCD18AGVaAFsIJOAoAFcbYBUAblCzADoAUTM+bgoYWmqAJQg4AAUnAVTC4vLKiGra7kaWttTB+oB5RvzlIuZQnX1DExhUq1t7cJhXdziIReh8GzSLGHcAxSyYAHNL6ADqJNEAVlEABlUNDnjElPTMnLyVN0qjU6hRBq0oO1Oh8vsk0hksrkpkDeiCBk0Ie1RhMIFMZnNdIcVidKEh8C5ssADrRjpR8MYPFR/sAyRRaPhaJSOXB8BRDhYIAV8L4IBYqXAfNSVu4qpQAqFYT8EczAQBJarAATcGwQWS4oz0aGaRXwv5IwEVYH9ABytHBwyNAF8WM7Zsw8LzEXk4b8vcAWOoOB6lWbveqqFqdXrjIb3hxA0GAOz0owwCgzDiu10sD3YCSEEhKy2o/oUAOcXP5kN+/DFvp1W32yEdYxpn3KpGEbCtihxhPYZPYbup9OcbMaSuSavM2s9esNDEOngptvTztDnt90IDrs9jPjoMEPOSKBNfCngQoOpCBDt0NUutoptYprjSaKZc9u81jcjrcfHdTx5X8033F0gA=)

**Reply by marchewek** [reply]
Only if you don't intend to modify the argument; otherwise you should use `ref`...
